client/state/data-layer/wpcom-http/pipeline/remove-duplicate-gets/index.js

```diff
@@ -0,0 +1,137 @@
+/**
+ * External dependencies
+ */
+import {
+	compact,
+	get,
+	head,
+	sortBy,
+	toPairs,
+	union,
+} from 'lodash';
+
+/**
+ * Prevent sending multiple identical GET requests
+ * while one is still transiting over the network
+ *
+ * Two requests are considered identical if they
+ * are both GET requests and share the same
+ * fundamental properties.
+ *
+ * @module state/data-layer/wpcom-http/optimizations/remove-duplicate-gets
+ */
+
+/** @type {Map} holds in-transit request keys */
+const requestQueue = new Map();
+
+/**
+ * Empties the duplication queue
+ *
+ * FOR TESTING ONLY!
+ */
+export const clearQueue = () => {
+	if ( 'undefined' !== typeof window ) {
+		throw new Error( '`clearQueue()` is not for use in production - only in testing!' );
+	}
+
+	requestQueue.clear();
+};
+
+/**
+ * Determines if a request object specifies the GET HTTP method
+ *
+ * @param {Object} request the HTTP request action
+ * @returns {Boolean} whether or not the method is GET
+ */
+const isGetRequest = request => 'GET' === get( request, 'method', '' ).toUpperCase();
+
+/**
+ * Generate a deterministic key for comparing request descriptions
+ *
+ * @param {String} path API endpoint path
+ * @param {String} apiNamespace used for endpoint versioning
+ * @param {String} apiVersion used for endpoint versioning
+ * @param {Object<String, *>} query GET query string
+ * @returns {String} unique key up to duplicate request descriptions
+ */
+export const buildKey = ( { path, apiNamespace, apiVersion, query } ) => JSON.stringify( [
+	path,
+	apiNamespace,
+	apiVersion,
+	sortBy( toPairs( query ), head ),
+] );
+
+/**
+ * Joins a responder action into a unique list of responder actions
+ *
+ * @param {Object<String, Object[]>} list existing responder actions
+ * @param {Object} item new responder action to add
+ * @returns {Object<String, Object[]>} union of existing list and new item
+ */
+export const addResponder = ( list, item ) => ( {
+	failures: union( list.failures, compact( [ item.onFailure ] ) ),
+	successes: union( list.successes, compact( [ item.onSuccess ] ) ),
+} );
+
+/**
+ * Prevents sending duplicate requests when one is
+ * already in transit over the network.
+ *
+ * @see applyDuplicateHandlers
+ *
+ * @type IngressProcessor
+ * @param {IngressData} ingressData
+ * @returns {IngressData}
+ */
+export const removeDuplicateGets = ingressData => {
+	const { nextRequest } = ingressData;
+
+	if ( ! isGetRequest( nextRequest ) ) {
+		return ingressData;
+	}
+
+	const key = buildKey( nextRequest );
+	const queued = requestQueue.get( key );
+	const request = addResponder( queued || { failures: [], successes: [] }, nextRequest );
```

Comment:
`addResponder()` and its tests show how we are merging the responder actions. if the same component makes the same request twice it will only dispatch that action once.

if we get into a logic error from calling one once or twice on response from an HTTP request then we have a problem at another level since it's always going to be the case that the network layer is unreliable.